lib/util/exec/hermit.ts

```diff
@@ -0,0 +1,90 @@
+import os from 'os';
+import fs from 'fs-extra';
+import upath from 'upath';
+import { GlobalConfig } from '../../config/global';
+import { logger } from '../../logger';
+import { rawExec } from './common';
+import type { RawExecOptions } from './types';
+
+export function isHermit(): boolean {
+  const { binarySource } = GlobalConfig.get();
+  return binarySource === 'hermit';
+}
+
+async function statFile(f: string): Promise<fs.Stats | undefined> {
+  let exists: fs.Stats | undefined = undefined;
+  try {
+    exists = await fs.stat(f);
+  } catch (e) {
+    // not doing anything when file not exists for errors
+  }
+
+  return exists;
+}
+
+async function statHermit(
+  defaultCwd: string,
+  parts: string[]
+): Promise<fs.Stats | undefined> {
+  const hermitForCwd = upath.join(...[defaultCwd, ...parts, 'bin', 'hermit']);
+  logger.trace({ hermitForCwd }, 'looking up hermit');
+  return await statFile(hermitForCwd);
+}
+
+export async function findHermitCwd(cwd: string): Promise<string> {
+  const defaultCwd = GlobalConfig.get('localDir') ?? '';
+  const parts = cwd.replace(defaultCwd, '').split(upath.sep);
+  let exists: fs.Stats | undefined = undefined;
+
+  // search the current relative path until reach the defaultCwd
+  while (parts.length > 0) {
+    exists = await statHermit(defaultCwd, parts);
+    // on file found. break out of the loop
+    if (exists !== undefined) {
+      break;
+    }
+    // otherwise, continue searching in parent directory
+    parts.pop();
+  }
+
+  // search in defaultCwd
+  if (exists === undefined) {
+    exists = await statHermit(defaultCwd, parts);
+  }
```

Comment:
Correct me if i'm wrong, 
aren't we calling `statHermit(DefaultCwd,parts)` in line 41? 
we call it until we find something,  and if we dont find something then parts will be empty, why call it again?
this can be done by doing `while(parts.length >= 0)` in line 40,

Reply:
the problem of having `while(parts.length >=0)` is that  when parts is popped empty and it wouldn't exit the loop unless conditionally checked inside.

however I can use the return value of `parts.pop()` to detect it. will update the code